grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/GrpcHeaderNames.java

```diff
@@ -34,6 +34,10 @@ public final class GrpcHeaderNames {
      * {@code "grpc-message"}.
      */
     public static final AsciiString GRPC_MESSAGE = HttpHeaderNames.of("grpc-message");
+    /**
+     * {@code "grpc-details"}.
+     */
+    public static final AsciiString GRPC_DETAILS = HttpHeaderNames.of("grpc-details");
```

Comment:
`grpc-details` is not an official gRPC header.
Should we prefix `ARMERIA_` to differentiate it from the official ones?
```suggestion
     * {@code "armeria.grpc.details-bin"}.
     */
    public static final AsciiString ARMERIA_GRPC_DETAILS = HttpHeaderNames.of("armeria.grpc.details-bin");
```

Reply:
Had a chat with the team and how about removing this customized header and using `GRPC_STATUS_DETAILS_BIN` above?
If we use this customized header, it won't be compatible with other clients. Only Armeria's client will understand it.
If we use `GRPC_STATUS_DETAILS_BIN` by the way, it will be compatible with other clients.

There are two problems though. `GRPC_STATUS_DETAILS_BIN` is not an official spec.
https://github.com/grpc/grpc/issues/24007
However, gRPC implementations use the header so I think it will become a spec one day.
https://github.com/grpc/grpc-java/blob/d4fa0ecc07495097453b0a2848765f076b9e714c/protobuf/src/main/java/io/grpc/protobuf/StatusProto.java#L114
https://github.com/grpc/grpc-go/blob/ea86bf74972b469b49cfbf6ab10258f478358036/internal/transport/handler_server.go#L212

The second problem is that we need to create the `io.grpc.Status` instance that the `grpc-protocol` module does not have the dependency.
Because `GRPC_STATUS_DETAILS_BIN` not only contains details but also contains the code and the message, we need the instance.
To solve this issue, we might want to define our own class:
```java
class ArmeriaStatus {
    int code;
    String message;
    byte[] details;
}
```
and serialize and deserialize it. (If creating instances become a problem, we might just serialize it without creating one.)

Despite these issues, I think we need to use `GRPC_STATUS_DETAILS_BIN` because I believe that supporting compatibility with other clients are the most important part.
Users who are migrating from other gRPC services to Armeria services won't be surprized by not support the detail headers.

Reply:
Please check out https://developers.google.com/protocol-buffers/docs/encoding to learn how to encode data into bytes.
It is a bit tricky but it is worth trying. 😄

Reply:
I thought about using the `GRPC_STATUS_DETAILS_BIN` and found that it also contains the code and the message, and I found a util (inside grpc module) and we don't want to rely on that module in grpc-protocol module so I didn't implement this way.

But since you raised this I will dig deeper to see whether I can encode/decode the `GRPC_STATUS_DETAILS_BIN` without using their util class

Reply:
> But since you raised this I will dig deeper to see whether I can encode/decode the GRPC_STATUS_DETAILS_BIN without using their util class

Oops, I missed this comment. 😅 
Please let us know if you get in trouble implementing it.

Reply:
> The second problem is that GRPC_STATUS_DETAILS_BIN not only contains details but also contains the code and the message. We might add StatusUtil.encode() and StatusUtil.decode() to solve the problem:

I think by design, the user of `grpc-protocol` and `AbstractUnaryGrpcService` is expected to serialize and deserialize byte arrays to the correct type. I think this is apparent in the `handleMessage` method

https://github.com/line/armeria/blob/c6b89064dfadfb985845876da86692e01e8bf1ec/grpc-protocol/src/main/java/com/linecorp/armeria/server/grpc/protocol/AbstractUnaryGrpcService.java#L43

Rather than manually encoding protobuf objects to byte arrays, what do you think of simply renaming the constructor for `ArmeriaStatusException` as follows?

As-Is:
```java
public ArmeriaStatusException(int code, @Nullable String message, @Nullable byte[] details)
```

To-Be:
```java
public ArmeriaStatusException(int code, @Nullable String message, @Nullable byte[] grpcStatusDetailsBin)
```

This way, it is more clear what the field means and delegates the responsibility of serializing the field to the user - which I assume he/she can do easily if a `protobuf` dependency is available.

Alternatively, I vote that we just add the `protobuf` dependency to the `grpc-protocol` module (I don't see why not!)

Reply:
Once this issue is settled, I think just modifying to use `grpc-status-details-bin` instead of `armeria.grpc.details-bin` will be enough for merging.

Reply:
> Alternatively, I vote that we just add the protobuf dependency to the grpc-protocol module (I don't see why not!)

IIRC, some users use `UnaryGrpcClient` to avoid protobuf and gRPC dependency, which is incompatible with their codebase.

Reply:
Yeah, we want to avoid adding the protobuf dependency. The first options looks good. Thanks, @jrhee17 for handling it.